libmamba/tests/src/util/test_url.cpp

```diff
@@ -0,0 +1,432 @@
+            CHECK_THROWS_AS(url.set_port("not-a-number"), std::invalid_argument);
+        }
+
+        SUBCASE("Encoding")
+        {
+            URL url{};
+            url.set_user("micro@mamba.pm", URL::Encode::yes);
+            CHECK_EQ(url.user(URL::Decode::no), "micro%40mamba.pm");
+            CHECK_EQ(url.user(URL::Decode::yes), "micro@mamba.pm");
+            url.set_password(R"(#!$&'"ab23)", URL::Encode::yes);
+            CHECK_EQ(url.password(URL::Decode::no), "%23%21%24%26%27%22ab23");
+            CHECK_EQ(url.password(URL::Decode::yes), R"(#!$&'"ab23)");
+            url.set_host("micro#mamba.org", URL::Encode::yes);
+            CHECK_EQ(url.host(URL::Decode::no), "micro%23mamba.org");
+            CHECK_EQ(url.host(URL::Decode::yes), "micro#mamba.org");
+        }
+    }
+
+    TEST_CASE("parse")
+    {
+        SUBCASE("Empty")
+        {
+            const URL url = URL::parse("");
+            CHECK_EQ(url.scheme(), URL::https);
+            CHECK_EQ(url.host(), URL::localhost);
+            CHECK_EQ(url.path(), "/");
+            CHECK_EQ(url.pretty_path(), "/");
+            CHECK_EQ(url.user(), "");
+            CHECK_EQ(url.password(), "");
+            CHECK_EQ(url.port(), "");
+            CHECK_EQ(url.query(), "");
+            CHECK_EQ(url.fragment(), "");
+        }
+
+        SUBCASE("mamba.org")
+        {
+            const URL url = URL::parse("mamba.org");
+            CHECK_EQ(url.scheme(), URL::https);
+            CHECK_EQ(url.host(), "mamba.org");
+            CHECK_EQ(url.path(), "/");
+            CHECK_EQ(url.pretty_path(), "/");
+            CHECK_EQ(url.user(), "");
+            CHECK_EQ(url.password(), "");
+            CHECK_EQ(url.port(), "");
+            CHECK_EQ(url.query(), "");
+            CHECK_EQ(url.fragment(), "");
+        }
+
+        SUBCASE("http://mamba.org")
+        {
+            const URL url = URL::parse("http://mamba.org");
+            CHECK_EQ(url.scheme(), "http");
+            CHECK_EQ(url.host(), "mamba.org");
+            CHECK_EQ(url.path(), "/");
+            CHECK_EQ(url.pretty_path(), "/");
+            CHECK_EQ(url.user(), "");
+            CHECK_EQ(url.password(), "");
+            CHECK_EQ(url.port(), "");
+            CHECK_EQ(url.query(), "");
+            CHECK_EQ(url.fragment(), "");
+        }
+
+        SUBCASE("s3://userx123:üúßsajd@mamba.org")
+        {
+            const URL url = URL::parse("s3://userx123:üúßsajd@mamba.org");
+            CHECK_EQ(url.scheme(), "s3");
+            CHECK_EQ(url.host(), "mamba.org");
+            CHECK_EQ(url.path(), "/");
+            CHECK_EQ(url.pretty_path(), "/");
+            CHECK_EQ(url.user(), "userx123");
+            CHECK_EQ(url.password(), "üúßsajd");
+            CHECK_EQ(url.port(), "");
+            CHECK_EQ(url.query(), "");
+            CHECK_EQ(url.fragment(), "");
+        }
+
+        SUBCASE("http://user%40email.com:test@localhost:8000")
```

Comment:
Maybe add a test for `http://user@email.com:test@localhost:8000`?

Reply:
It will be a parsing error. Adding a negative test.